src/Features/VisualBasic/Portable/CodeFixes/RemoveUnnecessaryCast/RemoveUnnecessaryCastCodeFixProvider.vb

```diff
@@ -6,181 +6,156 @@ Imports System.Threading
 
-        Public NotOverridable Overrides Async Function RegisterCodeFixesAsync(context As CodeFixContext) As Task
-            Dim document = context.Document
-            Dim span = context.Span
-            Dim cancellationToken = context.CancellationToken
-
-            Dim root = Await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(False)
-            Dim model = DirectCast(Await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(False), SemanticModel)
-
-            Dim node = GetCastNode(root, model, span, cancellationToken)
-            If node Is Nothing Then
-                Return
-            End If
-
-            context.RegisterCodeFix(
-                New MyCodeAction(
-                    VBFeaturesResources.Remove_Unnecessary_Cast,
-                    Function(c) RemoveUnnecessaryCastAsync(document, node, c)),
+        Public Overrides Function RegisterCodeFixesAsync(context As CodeFixContext) As Task
+            context.RegisterCodeFix(New MyCodeAction(
+                FeaturesResources.Remove_Unnecessary_Cast,
+                Function(c) FixAsync(context.Document, context.Diagnostics.First(), c)),
                 context.Diagnostics)
+            Return SpecializedTasks.EmptyTask
         End Function
 
-        Private Shared Function GetCastNode(root As SyntaxNode, model As SemanticModel, span As TextSpan, cancellationToken As CancellationToken) As ExpressionSyntax
-            Dim token = root.FindToken(span.Start)
-            If Not token.Span.IntersectsWith(span) Then
-                Return Nothing
-            End If
-
-            Dim node = token.GetAncestors(Of ExpressionSyntax)() _
-                            .Where(Function(c) TypeOf c Is CastExpressionSyntax OrElse TypeOf c Is PredefinedCastExpressionSyntax) _
-                            .FirstOrDefault(Function(c) c.Span.IntersectsWith(span) AndAlso IsUnnecessaryCast(c, model, cancellationToken))
-            Return node
-        End Function
-
-        Private Shared Function IsUnnecessaryCast(node As ExpressionSyntax, model As SemanticModel, cancellationToken As CancellationToken) As Boolean
-            Dim castExpression = TryCast(node, CastExpressionSyntax)
-            If castExpression IsNot Nothing Then
-                Return castExpression.IsUnnecessaryCast(model, assumeCallKeyword:=True, cancellationToken:=cancellationToken)
-            End If
-
-            Dim predefinedCastExpression = TryCast(node, PredefinedCastExpressionSyntax)
-            If predefinedCastExpression IsNot Nothing Then
-                Return predefinedCastExpression.IsUnnecessaryCast(model, assumeCallKeyword:=True, cancellationToken:=cancellationToken)
-            End If
-
-            Return False
+        Protected Overrides Async Function FixAllAsync(
+            document As Document, diagnostics As ImmutableArray(Of Diagnostic),
+            editor As SyntaxEditor, cancellationToken As CancellationToken) As Task
+
```

Comment:
note: VB and C# deviated sufficiently in their fix impl that i did not feel like it was worthwhile to unify on a base type.